src/target/riscv/riscv.c

```diff
@@ -1456,23 +1456,35 @@ static int old_or_new_riscv_poll(struct target *target)
 		return riscv_openocd_poll(target);
 }
 
+static struct reg *get_reg_cache_entry(struct target *target,
+		unsigned int number)
+{
+	assert(target->reg_cache);
+	assert(target->reg_cache->reg_list);
+	assert(number < target->reg_cache->num_regs);
+	return &target->reg_cache->reg_list[number];
+}
+
 int riscv_flush_registers(struct target *target)
 {
 	RISCV_INFO(r);
 
 	if (!target->reg_cache)
 		return ERROR_OK;
 
-	LOG_DEBUG("[%s]", target_name(target));
+	LOG_TARGET_DEBUG(target, "");
 
-	for (uint32_t number = 0; number < target->reg_cache->num_regs; number++) {
-		struct reg *reg = &target->reg_cache->reg_list[number];
+	/* Since writing a register can make some GPRs dirty (e.g. S0, S1),
+	 * registers should be flushed in reverse order, so GPRs are flushed last.
+	 */
```

Comment:
Since this is an important piece of information, I've tried to make it even more clear:

```suggestion
	/* Writing non-GPR registers may require progbuf execution, and some GPRs may 
	 * become dirty in the process (e.g. S0, S1). For that reason, flush registers in 
	 * reverse order, so that GPRs are flushed last. */
```

Reply:
Addressed.